#38 level unlocking system

joshjwilli opened the pull request:
- Implemented level unlocking as described in the issue
- Clicking A at the main menu resets playerprefs which is what saves what levels are unlocked (PlayerPrefs.DeleteAll(); can be added to the exit button or really whatever we want to reset the level unlocking system in the future.)
- Fixed the build order for scenes (the level unlocking system that I implemented relies on this order)
- Main menu, level select, and the two scenes we have should work together as intended
- Also removed a redundant exit button script so now both exit buttons (main menu and pause menu) should utilize the same exit button script

joshjwilli committed:
Implemented A "2nd level" to prove the current system works

joshjwilli committed:
Implemented issue

- Level unlocking works and should be scalable also implemented as described
- Fixed the build order for scenes
- Got the main menu, level select, and levels working properly together

joshjwilli committed:
Minor format improvements

joshjwilli committed:
Removed redundant exit button scripts

Removed redundant exit button scripts and just implemented one script for both current exit buttons to use

adenprince commented on Assets/Scripts/DeleterPlayerPrefs.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class DeleterPlayerPrefs : MonoBehaviour
```
I think you should name this class `DebugPlayerPrefsDeleter` (or something like that) to indicate that this script should not be used in a released build, but you can leave the name as is if you want.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -3,11 +3,35 @@
 
 public class LevelEndTriggerController : MonoBehaviour
 {
+    public int nextSceneLoad,
```
I think index should be included in the name of this variable (like `nextLevelSceneIndex`).

adenprince replied:
`[SerializeField] private` should be used instead of `public` here (also for `lvlButtons` in `LevelSelection.cs`). It's good to restrict variable visibility so that there are fewer potential possibilities to consider when looking for variable usage.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -3,11 +3,35 @@
+                SceneManager.LoadScene(nextSceneLoad);
+
+                // Setting Int for Index
+                if (nextSceneLoad > PlayerPrefs.GetInt("levelAt"))
```
I think `highestLevelUnlocked` (or something similar) would be clearer than `levelAt`.

joshjwilli replied:
I believe that's a default variable in PlayerPrefs, unsure if that's changeable.

adenprince replied:
You can use any key string with `PlayerPrefs`.

adenprince commented on Assets/Scripts/LevelSelection.cs:
```diff
@@ -0,0 +1,23 @@
+using System.Collections;
+using System.Collections.Generic;
```
These unused using directives should be removed. (There are also unused using directives in `LevelSelection.cs`.)

adenprince commented on Assets/Scripts/LevelSelection.cs:
```diff
@@ -0,0 +1,23 @@
+
+        for (int i = 0; i < lvlButtons.Length; i++)
+        {
+            if ((i + 2) > levelAt)
```
Here:
- `i` is level number (for the current level button) - 1.
- `levelAt` is the highest level number unlocked.

It would be simpler to make `i` go from `1` to `lvlButtons.Length` instead of `0` to `lvlButtons.Length - 1`. Then, the `if` statement becomes `if (i + 1 > levelAt)`.

joshjwilli replied:
There is no - 1 after lvlButtons.Length in the current code, I'm confused. I'm fairly certain the code is i + 2 because the lvlButtons array has the lvls starting at index 0 while the build order .GetInt("levelAt") would have the lvls starting at index 2 due to main menu being 0 and level selection being 1.

adenprince replied:
I was saying that `i` could iterate through level numbers (1-10) instead of level button indexes (0-9) so that the condition relates level numbers to level scene build indexes, but that's not really any better because then `lvlButtons[i - 1]` would need to be used.
I think it's better to just make level number correspond directly with level scene build index (as discussed [here](https://github.com/SIG-Game/VampireRun/pull/39#discussion_r1194446312)) and then use `i + 1` in the condition.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -3,11 +3,35 @@
 public class LevelEndTriggerController : MonoBehaviour
 {
+    public int nextSceneLoad,
+               lastLevel;
```
I think `lastLevel` should be a private constant (`private const int lastLevel = 2;`). That's simpler than making it an instance variable.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -3,11 +3,35 @@
+    public int nextSceneLoad,
+               lastLevel;
+
+    void Start()
```
I usually specify a `private` access modifier on methods instead of omitting it, but the visibility is going to default to `private` anyway. It's mostly personal preference, but I think we should be consistent by including `private`.

adenprince commented on Assets/Scripts/LevelSelection.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public Button[] lvlButtons;
+
+    // Start is called before the first frame update
```
This comment can be removed. (The comments in `LevelEndTriggerController.cs` can also be removed.)

adenprince commented on Assets/Scripts/UI/Handlers/QuitButton.cs:
```diff
@@ -1,7 +1,10 @@
+{
+    public void QuitButton_OnClick()
+    {
+        Debug.Log("Quit\n");
```
`\n` is unneeded here.

adenprince commented on ProjectSettings/EditorBuildSettings.asset:
```diff
@@ -5,10 +5,19 @@ EditorBuildSettings:
+    guid: 30277d48b45e13d4a8eb2a89feddf8cd
   - enabled: 1
-    path: Assets/Scenes/SampleScene.unity
+    path: Assets/Scenes/Level1.unity
```
The code would be simpler if the level number corresponded directly to the level scene build index (meaning that the scene Level1 would have a build index of 1). I think it would be good to do that.

joshjwilli replied:
How would I do that without forcing the build index of either main menu or level selection to change everytime we add a new level?

adenprince replied:
You can't do that without forcing the build index of one of those scenes to change when a level is added. Scenes can be loaded by scene name, so the scene with the highest build index can be loaded by scene name instead of build index.

adenprince commented:
It looks like you built the game inside of your repository clone and then pushed the build. You should remove the VampireGameRun folder because there doesn't need to be a build in the source code repository.

adenprince reviewed (changes requested):
Looks good. I requested some small changes.

joshjwilli committed:
Implemented Most Changes

Should have deleted the build folder from the files, unsure if just deleting the folder does the job.

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -3,11 +3,35 @@
+    [SerializeField]
+    private int indexForNextScene;
+
+    [SerializeField]
```
It seems like `[SerializeField]` doesn't do anything for `const` variables, so I think it can be removed.

adenprince commented on Assets/Scripts/UI/Handlers/LoadSceneButton.cs:
```diff
@@ -1,10 +1,20 @@
+        }
+        catch
+        {
+            Debug.Log("Scene has not been implemented yet.");
```
I think it's better to use `$"Scene {nameOfSceneToLoad} not found."` or something like that so that more information is provided about the scene name that failed to load. Also, having the scene name there would be helpful in determining if `nameOfSceneToLoad` is wrong.

adenprince commented on Assets/Scripts/LevelSelection.cs:
```diff
@@ -0,0 +1,21 @@
+    [SerializeField]
+    private Button[] lvlButtons;
+
+    void Start()
```
This can be `private void Start()`. It doesn't really matter to me, though.

adenprince commented on Assets/Scripts/DebugPlayerPrefsDeleter.cs:
```diff
@@ -0,0 +1,13 @@
+public class DebugPlayerPrefsDeleter : MonoBehaviour
+{
+    // Update is called once per frame
+    void Update()
```
This can be `private void Update()`, and the comment here can be removed.

adenprince commented on Assets/Scripts/DebugPlayerPrefsDeleter.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        if (Input.GetKeyDown(KeyCode.A))
+        {
+            PlayerPrefs.DeleteAll();
```
I think it would be good to log that PlayerPrefs were deleted here in case someone working on this project doesn't know about this script and presses the A key while playing `StartMenu.unity`.

adenprince commented on Assets/Scripts/LevelSelection.cs:
```diff
@@ -0,0 +1,23 @@
+
+        for (int i = 0; i < lvlButtons.Length; i++)
+        {
+            if ((i + 2) > levelAt)
```
There are extra parentheses around `i + 2`. You can keep them if you want to.

adenprince reviewed (changes requested):
Looks good. I just clarified a few things from my previous review and left comments for some small changes.

joshjwilli committed:
Implemented more change requests and reorganized folders

joshjwilli committed:
Made a levels folder in scenes

adenprince commented on Assets/Scripts/Game Controllers/LevelEndTriggerController.cs:
```diff
@@ -3,11 +3,34 @@
+        if (collision.CompareTag("Player"))
         {
-            SceneManager.LoadScene("LevelSelect");
+            if ((SceneManager.GetActiveScene().buildIndex - 1) == lastLevel)
```
This condition should be `SceneManager.GetActiveScene().buildIndex == lastLevel` because of the level scene build index change.

adenprince reviewed (changes requested):
Thank you for making the requested changes. I just left a comment about one small fix.

joshjwilli committed:
Fixed bug relating to change of build order

joshjwilli committed:
Merge remote-tracking branch 'origin/premain' into #38_level_unlocking_system

adenprince reviewed (changes requested):
Can you remove the Background GameObjects from both level scenes and add an instance of the BackgroundPack prefab to those scenes? Right now, there's an exception caused by the Background GameObjects that occurs when starting one of those scenes.

joshjwilli committed:
Testing

joshjwilli committed:
Revert "Merge remote-tracking branch 'origin/premain' into #38_level_unlocking_system"

This reverts commit 7e1cdd8b517ca672551cde1fec00bf88865e022e, reversing
changes made to f9dd8437dfdf60bd01dcad6deed197e2ca674810.

joshjwilli committed:
Revert "Fixed bug relating to change of build order"

This reverts commit f9dd8437dfdf60bd01dcad6deed197e2ca674810.

joshjwilli committed:
Revert "Revert "Fixed bug relating to change of build order""

This reverts commit 098181717fdb77cfe597fa4cb82442d79ea3debe.

joshjwilli committed:
Revert "Testing"

This reverts commit 69e202ba76f5d606ae8d8db89abd1ef64873d9f1.

joshjwilli committed:
Current State

joshjwilli committed:
Fixed merged conflicts

joshjwilli committed:
Revert "Revert "Merge remote-tracking branch 'origin/premain' into #38_level_unlocking_system""

This reverts commit bce1a0a685668730a54be39ae05fd13066f229af.

joshjwilli committed:
Merged premain into branch actually this time

adenprince reviewed (approved):
Looks good. Thank you.
